core/psh/dmesg/dmesg.c

```diff
@@ -0,0 +1,74 @@
+/*
+ * Phoenix-RTOS
+ *
+ * dmesg - reads kernel ring buffer
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Maciej Purski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "../psh.h"
+
+
+void psh_dmesginfo(void)
+{
+	printf("read kernel ring buffer");
+}
+
+
+static void psh_dmesg_help(const char *prog)
+{
+	printf("Usage: %s [options] [files]\n", prog);
+	printf("  -h:  shows this help message\n");
+}
+
+
+int psh_dmesg(int argc, char **argv)
+{
+	char buf[256];
+	int ret;
+	int c, fd;
+
+	while ((c = getopt(argc, argv, "h")) != -1) {
+		switch (c) {
+		case 'h':
+		default:
+			psh_dmesg_help(argv[0]);
+			return EOK;
+		}
+	}
+
+	if ((fd = open("/dev/kmsg", O_RDONLY | O_NONBLOCK)) < 0) {
+		printf("dmesg: Fail to open /dev/kmsg\n");
+		return 1;
+	}
+
+	while (1) {
+		ret = read(fd, buf, sizeof(buf));
+		if (ret < 0) {
+			close(fd);
+			break;
+		}
+		write(STDOUT_FILENO, buf, ret);
```

Comment:
buff[sizeof(buff) - 1] = '\0';
printf("%s", buf);

So we won't send '\0' to the stdout. We should be guaranteed to not get more than one line from kmsg from one read

Reply:
Also while errno == -EINTR